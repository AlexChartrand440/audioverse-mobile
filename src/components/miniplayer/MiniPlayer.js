import React from 'react'
import PropTypes from 'prop-types'
import { View, ActivityIndicator, Platform, StyleSheet } from 'react-native'
import TrackPlayer from 'react-native-track-player'

import ListItem from 'src/components/list/ListItem'
import ImageButton from 'src/components/buttons/ImageButton'
import ProgressBarMini from 'src/components/progressbar/ProgressBarMini'
import iconPlay from 'assets/ic_play.png'
import iconPause from 'assets/pause.png'

const MiniPlayer = ({ navigation, track, state, actions}) => {

  if (!track || state == TrackPlayer.STATE_NONE || state == TrackPlayer.STATE_STOPPED) {
    return <View />
  }

  const handlePress = () => { navigation.navigate({ routeName: 'Player' }) }

  const rightElement = state === TrackPlayer.STATE_BUFFERING ?
    <ActivityIndicator size="large" /> :
    <ImageButton
      source={state == TrackPlayer.STATE_PLAYING ? iconPause : iconPlay}
      imageStyle={styles.playPause}
      onPress={actions.playPause}
    />

  return (
    <View style={styles.container}>
      <ListItem
        avatar={{source: track.artwork}}
        title={track.title}
        subtitle={track.artist}
        onPress={handlePress}
        rightElement={rightElement}
        style={{backgroundColor: 'transparent'}}
      />
      <ProgressBarMini />
    </View>
  )

}

const styles = StyleSheet.create({
  container: {
    shadowOpacity: 0.75,
    shadowRadius: 5,
    shadowColor: '#000000',
    shadowOffset: { height: 0, width: 0 },
    elevation: 2,
    backgroundColor: '#E0E0E0CC'
  },
  playPause: {
    height: 42,
    width: 42,
    tintColor: '#000000'
  },
  bar: {
    height: 2,
    backgroundColor: '#03A9F4'
  },
})

MiniPlayer.propTypes = {
  navigation: PropTypes.object.isRequired,
  state: Platform.OS == 'ios' ? PropTypes.string : PropTypes.number,
  track: PropTypes.object,
  actions: PropTypes.shape({
    playPause: PropTypes.func.isRequired
  })
}

export default MiniPlayer
