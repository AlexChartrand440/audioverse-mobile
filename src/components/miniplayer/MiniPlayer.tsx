import React, { useState } from 'react';
import { ActivityIndicator, Platform, StyleSheet, View } from 'react-native';
import { ListItem } from 'react-native-elements';
import {
	Event as PlayerEvent,
	State as PlayerState,
	Track,
	usePlaybackState,
	useTrackPlayerEvents,
} from 'react-native-track-player';
import { NavigationInjectedProps } from 'react-navigation';

import iconPlay from '../../../assets/ic_play.png';
import iconPause from '../../../assets/pause.png';
import I18n from '../../../locales';
import data from '../../constants/prompts';
import ImageButton from '../buttons/ImageButton';
import ProgressBarMini from '../progressbar/ProgressBarMini';

// eslint-disable-next-line @typescript-eslint/no-var-requires
const MarqueeText: any = require('react-native-marquee').default;

interface Props extends NavigationInjectedProps {
	track: Track | undefined;
	actions: {
		playPause: () => void;
	};
}

const styles = StyleSheet.create({
	container: {
		shadowOpacity: 0.75,
		shadowRadius: 5,
		shadowColor: '#000000',
		shadowOffset: { height: 0, width: 0 },
		elevation: 2,
		backgroundColor: '#E0E0E0',
	},
	playPause: {
		height: 42,
		width: 42,
		tintColor: '#000000',
	},
	title: {
		fontSize: Platform.OS === 'ios' ? 17 : 16,
	},
});

const MiniPlayer: React.FC<Props> = ({ navigation, track, actions }) => {
	const playbackState = usePlaybackState();
	const [loading, setLoading] = useState(false);
	// on iOS the Media Player library that we are using does not
	// enter the loading state until it has loaded/buffered the mp3 file, however
	// in order to let the user know that the file is being loaded we are
	// using our own loading state, it is set to true when the a track is changed
	// and is set to false when the playback-state change.
	useTrackPlayerEvents([PlayerEvent.PlaybackTrackChanged], (data) => {
		if (Platform.OS === 'ios' && data.nextTrack) {
			setLoading(true); // show activity indicator
		}
	});

	useTrackPlayerEvents([PlayerEvent.PlaybackState], (event) => {
		if (event.state === PlayerState.Playing && loading) {
			setLoading(false); // hide activity indicator
		}
	});

	if (!track) {
		return null;
	}

	const handlePress = () => {
		navigation.navigate({ routeName: 'Player' });
	};
	const isLoading = loading || playbackState === PlayerState.Buffering;
	const rightElement = isLoading ? (
		<ActivityIndicator size="large" color="black" />
	) : (
		<ImageButton
			source={playbackState === PlayerState.Playing ? iconPause : iconPlay}
			imageStyle={styles.playPause}
			onPress={actions.playPause}
			accessibilityLabel={I18n.t(playbackState === PlayerState.Playing ? 'pause' : 'play')}
		/>
	);

	return (
		<View style={styles.container} testID="mini-player">
			<ListItem
				leftAvatar={{
					source:
						track.artwork && track.artwork.toString().startsWith('http')
							? { uri: track.artwork }
							: (track.artwork as any),
				}}
				title={
					<MarqueeText
						marqueeOnStart
						duration={3500}
						loop
						style={styles.title}
						accessibilityHint={I18n.t('maximize_player')}>
						{track.title}
					</MarqueeText>
				}
				subtitle={track.artist}
				subtitleProps={{ numberOfLines: 1 }}
				onPress={handlePress}
				rightElement={rightElement}
				containerStyle={{ backgroundColor: 'transparent' }}
				underlayColor="#E0E0E0"
			/>
			<ProgressBarMini />
		</View>
	);
};

export default MiniPlayer;
