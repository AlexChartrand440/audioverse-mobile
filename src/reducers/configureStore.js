import { createStore, applyMiddleware } from 'redux'
import createSagaMiddleware from 'redux-saga'
// use next import to use react-navigation with redux
// import { middleware as reactNavigationMiddleware } from 'src/utils/react-navigation-redux'
import { persistStore } from 'redux-persist'

import reducer from './index'
import mySaga from 'src/sagas'

export default () => {
  // create the saga middleware
  const sagaMiddleware = createSagaMiddleware()

  // use this to use react-navigation no redux
  const store = createStore(reducer, applyMiddleware(sagaMiddleware))
  // use this to use react-navigation with redux
  // const store = createStore(reducer, applyMiddleware(sagaMiddleware, reactNavigationMiddleware))

  // persistor
  const persistor = persistStore(store)

  // run the saga
  sagaMiddleware.run(mySaga)

  return { store, persistor }
}
