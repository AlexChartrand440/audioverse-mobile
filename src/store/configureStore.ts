import { applyMiddleware, createStore } from 'redux'
import { persistStore } from 'redux-persist'
import createSagaMiddleware from 'redux-saga'

import mySaga from '../sagas'

import reducer from './index'

export default () => {
  // create the saga middleware
  const sagaMiddleware = createSagaMiddleware()

  const store = createStore(reducer, applyMiddleware(sagaMiddleware))

  // persistor
  const persistor = persistStore(store)

  // run the saga
  sagaMiddleware.run(mySaga)

  return { store, persistor }
}
