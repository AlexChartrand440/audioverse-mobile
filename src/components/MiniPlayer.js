import React from 'react'
import PropTypes from 'prop-types'
import { View, Platform, StyleSheet } from 'react-native'
import { connect } from 'react-redux'
import TrackPlayer from 'react-native-track-player'

import ListItem from '../components/ListItem'
import ImageButton from '../components/ImageButton'
import ProgressBarMini from '../components/ProgressBarMini'
import { playPause } from '../actions'
import { getCurrentTrack, getPlaybackState } from '../reducers/selectors'
import iconPlay from '../../assets/ic_play.png'
import iconPause from '../../assets/pause.png'

const MiniPlayer = ({ onPressMetaData, track, state, playPause}) => {

  if (!track || state == TrackPlayer.STATE_NONE || state == TrackPlayer.STATE_STOPPED) {
    return <View />
  }

  const rightElement = <ImageButton
    source={state == TrackPlayer.STATE_PAUSED ? iconPlay : iconPause}
    imageStyle={styles.playPause}
    onPress={playPause}
  />

  return (
    <View style={styles.container}>
      <ListItem
        avatar={{source: track.artwork}}
        title={track.title}
        subtitle={track.artist}
        onPress={onPressMetaData}
        rightElement={rightElement}
        style={{backgroundColor: 'transparent'}}
      />
      <ProgressBarMini />
    </View>
  )

}

const styles = StyleSheet.create({
  container: {
    shadowOpacity: 0.75,
    shadowRadius: 5,
    shadowColor: '#000000',
    shadowOffset: { height: 0, width: 0 },
    elevation: 2,
    backgroundColor: '#E0E0E0CC'
  },
  playPause: {
    height: 42,
    width: 42,
    tintColor: '#000000'
  },
  bar: {
    height: 2,
    backgroundColor: '#03A9F4'
  },
})

MiniPlayer.propTypes = {
  onPressMetaData: PropTypes.func.isRequired,
  state: Platform.OS == 'ios' ? PropTypes.string : PropTypes.number,
  track: PropTypes.object,
  playPause: PropTypes.func.isRequired
}

const  mapStateToProps = state => ({
  state: getPlaybackState(state),
  track: getCurrentTrack(state)
})

const mapDispatchToProps = dispatch => ({
  playPause: () => dispatch(playPause())
})

export default connect(mapStateToProps, mapDispatchToProps)(MiniPlayer)
