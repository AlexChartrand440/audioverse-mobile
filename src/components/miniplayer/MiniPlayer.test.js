import React from 'react'
import MiniPlayer from './'

import { shallow } from 'enzyme'
import configureStore from 'redux-mock-store'

const middlewares = []
const mockStore = configureStore(middlewares)

describe('MiniPlayer component', () => {
  it('renders empty component when no track', () => {
    const initialState = {
      playback: { state: '' }
    }
    const wrapper = shallow(<MiniPlayer onPressMetaData={jest.fn()} />, { context: { store: mockStore(initialState) } })
    expect(wrapper.dive()).toMatchSnapshot()
  })

  it('renders correctly when there is a track', () => {
    const initialState = {
      playback: { state: '', currentTrack: {} }
    }
    const wrapper = shallow(<MiniPlayer onPressMetaData={jest.fn()} />, { context: { store: mockStore(initialState) } })
    expect(wrapper.dive()).toMatchSnapshot()
  })
})
