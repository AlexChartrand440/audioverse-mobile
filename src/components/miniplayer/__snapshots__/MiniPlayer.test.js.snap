// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MiniPlayer component renders correctly when there is a track 1`] = `<Component />`;

exports[`MiniPlayer component renders empty component when no track 1`] = `<Component />`;
